x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java

```diff
@@ -95,80 +99,105 @@ private FilterExec rewrite(FilterExec filterExec, EsQueryExec esQueryExec) {
         return filterExec;
     }
 
-    private void getPushableDistances(List<Alias> aliases, Map<NameId, StDistance> distances, Map<NameId, Alias> others) {
+    /**
+     * This version of the rewrite will try to split the filter into two parts, one that can be pushed down to the source and
+     * one that needs to be kept after the EVAL.
+     * For example:
+     * <pre>
+     *     FROM index
+     *     | EVAL distance = ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")), other = scale * 2
+     *     | WHERE distance &lt;= 10000 AND distance &gt; 5000 AND other &gt; 10
+     * </pre>
+     * Should be rewritten as:
+     * <pre>
+     *     FROM index
+     *     | WHERE ST_INTERSECTS(field, TO_GEOSHAPE("CIRCLE(0,0,10000)"))
+     *         AND ST_DISJOINT(field, TO_GEOSHAPE("CIRCLE(0,0,5000)"))
+     *     | EVAL distance = ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")), other = scale * 2
+     *     | WHERE other &gt; 10
+     * </pre>
+     */
+    private PhysicalPlan rewriteBySplittingFilter(FilterExec filterExec, EvalExec evalExec, EsQueryExec esQueryExec) {
+        // Find all pushable distance functions in the EVAL
+        Map<NameId, StDistance> distances = getPushableDistances(evalExec.fields());
+
+        // Don't do anything if there are no distances to push down
+        if (distances.isEmpty()) {
+            return filterExec;
+        }
+
+        // Process the EVAL to get all aliases that might be needed in the filter rewrite
+        AttributeMap.Builder<Attribute> aliasReplacedByBuilder = AttributeMap.builder();
+        evalExec.fields().forEach(alias -> {
+            if (alias.child() instanceof Attribute attr) {
+                aliasReplacedByBuilder.put(alias.toAttribute(), attr);
+            }
+        });
+        AttributeMap<Attribute> aliasReplacedBy = aliasReplacedByBuilder.build();
+
+        // First we split the filter into multiple AND'd expressions, and then we evaluate each individually for distance rewrites
+        List<Expression> pushable = new ArrayList<>();
+        List<Expression> nonPushable = new ArrayList<>();
+        for (Expression exp : splitAnd(filterExec.condition())) {
+            Expression resExp = exp.transformUp(ReferenceAttribute.class, r -> aliasReplacedBy.resolve(r, r));
+            // Find and rewrite any binary comparisons that involve a distance function and a literal
+            var rewritten = rewriteDistanceFilters(resExp, distances);
+            // If all pushable StDistance functions were found and re-written, we need to re-write the FILTER/EVAL combination
+            if (rewritten.equals(resExp) == false && conditionHasNoReferences(rewritten, distances)) {
```

Comment:
Isn't `conditionHasNoReferences(rewritten, distances)` always true and thus redundant if the rewrite was succesful?

Reply:
That fix originally fixed the conjunctive `WHERE distance < 50000 AND 0*distance < 40000`, but while the new split-filter should also fix that, I think we need this for disjunctive cases, like `WHERE distance < 50000 OR 0*distance < 40000`.